File: 18-12-2024_Final_Prices_With_a_Special_Discount_in_a_Shop.py
class Solution:
    def finalPrices(self, prices: List[int]) -> List[int]:
        ans = prices.copy()
        stack = deque()
        for i in range(len(prices)):
            while stack and prices[stack[-1]] >= prices[i]:
                ans[stack.pop()] -= prices[i]
            stack.append(i)
        return ans

File: 18-12-2024_Final_Prices_With_a_Special_Discount_in_a_Shop.cpp
class Solution {
public:
    vector<int> finalPrices(vector<int>& prices) {
        int n = prices.size();
        vector<int> ans = prices;
        stack<int> stack;
        for(int i=0; i<n; i++){
            while(!stack.empty() && prices[stack.top()] >= prices[i]){
                ans[stack.top()] -= prices[i];
                stack.pop();
            }
            stack.push(i);
        }
        return ans;
    }
};
